File: src/Nezam.Refahi.Web/wwwroot/js/vlz-table.js
(function() {
    'use strict';

    // Initialize all VlzTable instances on the page
    function initVlzTables() {
        document.querySelectorAll('table.vlz-data-table').forEach(initTable);
    }

    // Initialize a single table instance
    function initTable(tbl) {
        if (!tbl || tbl._vlzInitialized) return;
        tbl._vlzInitialized = true;

        const tableId = tbl.id;
        const wrapper = tbl.closest('.table-responsive');
        // Allow either data-show-filter-row or data-show-filter
        const showFilter = (tbl.dataset.showFilterRow === 'true') || (tbl.dataset.showFilter === 'true');
        const formId = tbl.dataset.filterFormId;
        const form = formId ? document.getElementById(formId) : null;
        const enableSorting = tbl.dataset.enableSorting !== 'false';
        const enablePagination = tbl.dataset.enablePagination !== 'false';
        const pageSize = parseInt(tbl.dataset.pageSize || '10', 10);
        const ajaxUrl = tbl.dataset.ajaxUrl;
        const method = (tbl.dataset.method || 'GET').toUpperCase();

        console.log(`[VlzTable] initTable: id=${tableId}, showFilter=${showFilter}, sorting=${enableSorting}, pagination=${enablePagination}, pageSize=${pageSize}`);

        // Columns config
        const inputs = Array.from(wrapper.querySelectorAll('input.vlz-column-config'));
        const columns = inputs.map(inp => {
            const col = {};
            Object.entries(inp.dataset).forEach(([k, v]) => {
                if (k === 'parentId') return;
                if (v === 'true') col[k] = true;
                else if (v === 'false') col[k] = false;
                else if (!isNaN(v) && v.trim() !== '') col[k] = Number(v);
                else col[k] = v;
            });
            return col;
        }).sort((a, b) => (a.order || 0) - (b.order || 0));

        // Build table header
        const thead = document.createElement('thead');
        const headerRow = document.createElement('tr');
        const filterRow = document.createElement('tr');
        
        // Ensure rows take full width
        headerRow.style.width = '100%';
        filterRow.style.width = '100%';
        
        // Add a class to the table for full-width rows
        tbl.classList.add('vlz-full-width-rows');

        columns.forEach(c => {
            const th = document.createElement('th');
            th.dataset.colName = c.name;
            if (c.cssClass) th.className = c.cssClass;
            if (c.width) th.style.width = c.width;
            if (c.visible === false) th.style.display = 'none';
            if (c.templateId) th.setAttribute('data-template-id', c.templateId);
            if (c.tooltip) th.title = c.tooltip;
            if (c.sortable) th.setAttribute('data-sortable', 'true');
            if (c.dataType) th.setAttribute('data-type', c.dataType);
            if (c.fixed) th.setAttribute('data-fixed', 'true');
            th.textContent = c.title;
            headerRow.appendChild(th);

            const thf = document.createElement('th');
            if (showFilter && c.filterable) {
                const inp = document.createElement('input');
                inp.type = c.filterType === 'number' ? 'number' : 'text';
                inp.name = c.name;
                inp.className = 'form-control form-control-sm';
                inp.placeholder = 'جستجو…';
                
                // Set minimum width for filter inputs
                inp.style.minWidth = '80px';
                inp.style.width = '100%';
                
                // If the column has a width, apply it to the filter cell too
                if (c.width) {
                    thf.style.width = c.width;
                }
                
                thf.appendChild(inp);
            }
            filterRow.appendChild(thf);
        });

        thead.appendChild(headerRow);
        if (showFilter) thead.appendChild(filterRow);
        if (tbl.tHead) tbl.replaceChild(thead, tbl.tHead);
        else tbl.insertBefore(thead, tbl.firstChild);

        tbl.vlzColumns = columns;
        const body = tbl.tBodies[0];
        let sortState = { column: null, direction: 'asc' };
        let activeFilters = {};
        let currentPage = 1;
        let filterTimeout;
        const filterDelay = 300;

        // Render rows function
        function renderRows(items) {
            console.log(`[VlzTable] renderRows: items.length=${items.length}`);
            body.innerHTML = '';
            if (!items || !items.length) {
                body.insertAdjacentHTML('beforeend',
                    `<tr style="width:100%"><td colspan="${columns.length}" class="text-center py-4">موردی یافت نشد.</td></tr>`);
                return;
            }
            items.forEach((item, idx) => {
                const tr = document.createElement('tr');
                tr.dataset.rowIndex = idx;
                tr.dataset.rowData = JSON.stringify(item);
                tr.style.width = '100%'; // Ensure data rows take full width
                columns.forEach(c => {
                    if (c.visible === false) return;
                    const td = document.createElement('td');
                    td.dataset.colName = c.name;
                    if (c.width) td.style.width = c.width;
                    if (c.cssClass) td.className = c.cssClass;
                    let val = getValueCaseInsensitive(item, c.name) ?? c.defaultValue ?? '';
                    td.dataset.value = val;
                    // Check if column has a template
                    if (c.templateId || c.templateContent) {
                        let templateHtml = '';
                        
                        // First priority: Use template from DOM if templateId is provided
                        if (c.templateId) {
                            const template = document.getElementById(c.templateId);
                            if (template) {
                                templateHtml = template.innerHTML;
                            } else {
                                console.warn(`Template with ID ${c.templateId} not found`);
                            }
                        }
                        // Second priority: Use inline template content if provided
                        else if (c.templateContent) {
                            templateHtml = c.templateContent;
                        }
                        
                        // If we have a template, process it
                        if (templateHtml) {
                            // Replace all {{property}} placeholders with actual values
                            Object.keys(item).forEach(key => {
                                const regex = new RegExp(`\{\{${key}\}\}`, 'g');
                                const value = item[key] !== undefined && item[key] !== null ? item[key] : '';
                                templateHtml = templateHtml.replace(regex, value);
                            });
                            
                            // Add data attributes to the cell for potential event handling
                            Object.keys(item).forEach(key => {
                                td.dataset[`row${key.charAt(0).toUpperCase() + key.slice(1)}`] = item[key] || '';
                            });
                            
                            td.innerHTML = templateHtml;
                        } else {
                            td.textContent = val || '';
                        }
                    } 
                    // Otherwise format based on data type
                    else {
                        switch ((c.dataType || '').toLowerCase()) {
                            case 'number':
                                td.textContent = c.format && !isNaN(+val)
                                    ? (+val).toLocaleString() : val;
                                break;
                            case 'price':
                                if (!isNaN(+val)) {
                                    td.dataset.rawValue = val;
                                    td.textContent = (+val).toLocaleString();
                                } else td.textContent = val;
                                break;
                            case 'date':
                                const d = new Date(val);
                                td.textContent = isNaN(d) ? val : d.toLocaleDateString();
                                break;
                            default:
                                td.textContent = val;
                        }
                    }
                    tr.appendChild(td);
                });
                body.appendChild(tr);
            });
        }

        // Helper: case-insensitive key lookup
        function getValueCaseInsensitive(obj, key) {
            if (obj[key] !== undefined) return obj[key];
            const lowerKey = key.toLowerCase();
            for (const k in obj) if (k.toLowerCase() === lowerKey) return obj[k];
            return undefined;
        }

        // Load data via AJAX
        async function loadData(page = 1) {
            currentPage = page;
            console.log(`[VlzTable] loadData: page=${page}, sort=${sortState.column}:${sortState.direction}, filters=${JSON.stringify(activeFilters)}`);
            const params = new URLSearchParams();
            // Filters
            Object.entries(activeFilters).forEach(([k, v]) => { if (v) params.set(k, v); });
            if (form) new FormData(form).forEach((v, k) => { if (v && !params.has(k)) params.set(k, v); });
            // Sorting
            if (sortState.column) {
                params.set('sortColumn', sortState.column);
                params.set('sortDirection', sortState.direction);
            }
            // Pagination
            params.set('pageSize', pageSize);
            params.set('page', page);

            // Update URL
            const base = ajaxUrl.split('?')[0];
            const requestUrl = `${base}?${params.toString()}`;
            window.history.replaceState(null, '', requestUrl);

            // Loading indicator
            body.innerHTML = `<tr style="width:100%"><td colspan="${columns.length}" class="text-center py-4">در حال بارگذاری…</td></tr>`;

            try {
                const headers = { 'X-Requested-With': 'XMLHttpRequest' };
                if (method !== 'GET') headers['Content-Type'] = 'application/json';
                const rsp = await fetch(requestUrl, { method, headers });
                if (!rsp.ok) throw new Error(`HTTP ${rsp.status}`);
                const json = await rsp.json();
                const items = json.items || json.data || [];
                renderRows(items);
                if (enablePagination && json.totalPages > 1) updatePagination(json.currentPage || page, json.totalPages, json.totalItems);
            } catch (e) {
                console.error('[VlzTable] loadData error:', e);
                body.innerHTML = `<tr><td colspan="${columns.length}" class="text-center py-4 text-danger">خطا در بارگذاری داده‌ها.</td></tr>`;
            }
        }

        // Pagination controls
        function updatePagination(current, total, totalItems) {
            let container = document.getElementById(`${tableId}-pagination`);
            if (!container) {
                container = document.createElement('div');
                container.id = `${tableId}-pagination`;
                container.className = 'vlz-pagination d-flex justify-content-between align-items-center mt-3';
                tbl.parentNode.insertBefore(container, tbl.nextSibling);
            }
            const showFrom = Math.min((current - 1) * pageSize + 1, totalItems);
            const showTo = Math.min(current * pageSize, totalItems);
            let html = `<div class="pagination-info">نمایش ${showFrom} تا ${showTo} از ${totalItems} مورد</div><ul class="pagination mb-0">`;
            // Previous
            html += `<li class="page-item ${current === 1 ? 'disabled' : ''}"><a class="page-link" href="#" data-page="${current-1}">&laquo;</a></li>`;
            const maxPages = 5;
            let start = Math.max(1, current - Math.floor(maxPages/2));
            let end = Math.min(total, start + maxPages - 1);
            if (end - start + 1 < maxPages && start > 1) start = Math.max(1, end - maxPages + 1);
            if (start > 1) { html += `<li class="page-item"><a class="page-link" href="#" data-page="1">1</a></li>`; if (start > 2) html += `<li class="page-item disabled"><span class="page-link">...</span></li>`; }
            for (let i = start; i <= end; i++) html += `<li class="page-item ${i === current ? 'active' : ''}"><a class="page-link" href="#" data-page="${i}">${i}</a></li>`;
            if (end < total) { if (end < total-1) html += `<li class="page-item disabled"><span class="page-link">...</span></li>`; html += `<li class="page-item"><a class="page-link" href="#" data-page="${total}">${total}</a></li>`; }
            html += `<li class="page-item ${current === total ? 'disabled' : ''}"><a class="page-link" href="#" data-page="${current+1}">&raquo;</a></li></ul>`;
            container.innerHTML = html;
            container.querySelectorAll('.page-link').forEach(link => {
                link.addEventListener('click', e => {
                    e.preventDefault();
                    const p = parseInt(e.currentTarget.dataset.page, 10);
                    if (!isNaN(p) && p > 0 && p <= total) loadData(p);
                });
            });
        }

        // Sorting setup
        if (enableSorting) {
            tbl.querySelectorAll('th[data-sortable]').forEach(th => {
                th.classList.add('sortable');
                th.addEventListener('click', () => {
                    const col = th.dataset.colName;
                    sortState.direction = (sortState.column === col && sortState.direction === 'asc') ? 'desc' : 'asc';
                    sortState.column = col;
                    console.log(`[VlzTable] sorting: ${col} ${sortState.direction}`);
                    tbl.querySelectorAll('th.sort-asc, th.sort-desc').forEach(x => x.classList.remove('sort-asc','sort-desc'));
                    th.classList.add(`sort-${sortState.direction}`);
                    loadData(1);
                });
            });
        }

        // Filtering inputs
        if (showFilter) {
            const headerInputs = Array.from(tbl.querySelectorAll('thead input'));
            const formInputs = form ? Array.from(form.querySelectorAll('input, select')) : [];
            const allInputs = headerInputs.concat(formInputs);
            allInputs.forEach(input => {
                if (input.name && input.value) activeFilters[input.name] = input.value;
                input.addEventListener('input', () => {
                    clearTimeout(filterTimeout);
                    filterTimeout = setTimeout(() => {
                        if (input.name) {
                            if (input.value) activeFilters[input.name] = input.value;
                            else delete activeFilters[input.name];
                        }
                        console.log(`[VlzTable] filter change: ${input.name}=${input.value}`);
                        loadData(1);
                    }, filterDelay);
                });
            });
            if (form) {
                form.addEventListener('submit', e => { e.preventDefault(); loadData(1); });
            }
        }

        // Initial data load
        loadData(1);
    }

    // Export to global scope
    window.VlzTable = { init: initVlzTables, initTable };
    document.addEventListener('DOMContentLoaded', initVlzTables);
})();

File: src/Nezam.Refahi.Web/TagHelpers/VlzTable.cs
using System.Text.Json.Serialization;
using Microsoft.AspNetCore.Razor.TagHelpers;

namespace Nezam.Refahi.Web.TagHelpers;

// ───────────────────────────────────────── V L Z   B A S E   T A B L E ───────────────────────────────────
public abstract class VlzBaseTableTagHelper : TagHelper
{
    // Table configuration properties
    [HtmlAttributeName("ajax-url")]        public string? AjaxUrl       { get; set; }
    [HtmlAttributeName("method")]          public string  Method        { get; set; } = "get";
    protected string NormalizedMethod => Method?.ToUpper() == "GET" || Method?.ToUpper() == "POST" ? Method.ToUpper() : "GET";
    [HtmlAttributeName("filter-form-id")]  public string? FilterFormId  { get; set; }
    [HtmlAttributeName("show-filter-row")] public bool    ShowFilterRow { get; set; }
    [HtmlAttributeName("table-id")]        public string? TableId       { get; set; }
    [HtmlAttributeName("page-size")]       public int     PageSize      { get; set; } = 10;
    [HtmlAttributeName("enable-sorting")]  public bool    EnableSorting { get; set; } = true;
    [HtmlAttributeName("enable-pagination")] public bool EnablePagination { get; set; } = true;
    [HtmlAttributeName("enable-column-selection")] public bool EnableColumnSelection { get; set; } = true;
    
    protected string GenerateTableId()
    {
        return TableId ?? $"vlz-table-{Guid.NewGuid():N}";
    }
}

// ───────────────────────────────────────── V L Z   C O L U M N ──────────────────────────────────────────
[HtmlTargetElement("vlz-table-column", ParentTag = "vlz-data-table")]
public sealed class VlzTableColumnTagHelper : TagHelper
{
    [HtmlAttributeName("name")]           public string  Name        { get; set; } = string.Empty;
    [HtmlAttributeName("title")]          public string  Title       { get; set; } = string.Empty;
    [HtmlAttributeName("class")]          public string? CssClass    { get; set; }
    [HtmlAttributeName("template")]       public string? TemplateId  { get; set; }
    [HtmlAttributeName("template-content")] public string? TemplateContent { get; set; }
    [HtmlAttributeName("filterable")]     public bool    Filterable  { get; set; }
    [HtmlAttributeName("filter-type")]    public string? FilterType  { get; set; }
    [HtmlAttributeName("sortable")]       public bool    Sortable    { get; set; } = true;
    [HtmlAttributeName("width")]          public string? Width       { get; set; }
    [HtmlAttributeName("visible")]        public bool    Visible     { get; set; } = true;
    [HtmlAttributeName("datatype")]       public string? DataType    { get; set; } = "text";
    [HtmlAttributeName("format")]         public string? Format      { get; set; }
    [HtmlAttributeName("searchable")]     public bool    Searchable  { get; set; } = true;
    [HtmlAttributeName("default-value")]  public string? DefaultValue { get; set; }
    [HtmlAttributeName("display-order")]  public int     DisplayOrder { get; set; } = 0;
    [HtmlAttributeName("tooltip")]        public string? Tooltip     { get; set; }
    [HtmlAttributeName("fixed")]          public bool    Fixed       { get; set; }

    [HtmlAttributeName("parent-id")]
    public string ParentId { get; set; } = string.Empty;

    public override async Task ProcessAsync(TagHelperContext context, TagHelperOutput output)
    {
        // Get the child content if available (for inline templates)
        string? childContent = null;
        
        // First priority: Check if there's actual content inside the tag
        if (output.Content.IsModified || !output.Content.IsEmptyOrWhiteSpace)
        {
            childContent = (await output.GetChildContentAsync()).GetContent();
        }
        // Second priority: Check if template-content attribute is provided
        else if (context.AllAttributes.ContainsName("template-content"))
        {
            childContent = TemplateContent;
        }
        
        var column = new VlzTableColumn
        {
            Name           = Name,
            Title          = Title,
            CssClass       = CssClass,
            TemplateId     = TemplateId,
            TemplateContent = childContent,
            Filterable     = Filterable,
            FilterType     = FilterType,
            Sortable       = Sortable,
            Width          = Width,
            Visible        = Visible,
            DataType       = DataType,
            Format         = Format,
            Searchable     = Searchable,
            DefaultValue   = DefaultValue,
            Order          = DisplayOrder,
            Tooltip        = Tooltip,
            Fixed          = Fixed
        };

        // Switch to an <input /> with self-closing mode
        output.TagName  = "input";
        output.TagMode  = TagMode.SelfClosing;
        output.Attributes.SetAttribute("type", "hidden");
        output.Attributes.SetAttribute("class", "vlz-column-config");
        output.Attributes.SetAttribute("data-parent-id", ParentId);
        output.Attributes.SetAttribute("data-column-name", Name);
        
        // Store template content in a data attribute if available
        if (!string.IsNullOrEmpty(childContent))
        {
            // Store the template content directly
            output.Attributes.SetAttribute("data-template-content", childContent);
        }

        // Reflect over the VlzTableColumn props and add each as a data-* attribute
        foreach (var prop in typeof(VlzTableColumn).GetProperties())
        {
            var rawValue = prop.GetValue(column, null);
            if (rawValue == null) 
                continue;

            string stringValue;

            // Booleans and numbers can use ToString(); strings we trim
            switch (rawValue)
            {
                case bool b:
                    stringValue = b.ToString().ToLowerInvariant(); 
                    break;
                case int i:
                    stringValue = i.ToString();
                    break;
                default:
                    stringValue = rawValue.ToString()!;
                    if (string.IsNullOrWhiteSpace(stringValue))
                        continue;
                    break;
            }

            // convert PascalCase to kebab-case for data- attribute name
            
            var kebab = global::System.Text.RegularExpressions.Regex.Replace(prop.Name, "([a-z])([A-Z])", "$1-$2")
                            .ToLowerInvariant();

            output.Attributes.SetAttribute($"data-{kebab}", stringValue);
        }
    }
}


// ───────────────────────────────────────── V L Z   D A T A   T A B L E ───────────────────────────────────
[HtmlTargetElement("vlz-data-table")]
public sealed class VlzDataTableTagHelper : VlzBaseTableTagHelper
{
    public override async Task ProcessAsync(TagHelperContext context, TagHelperOutput output)
    {
        var childHtml = (await output.GetChildContentAsync()).GetContent();
        var tableId   = GenerateTableId();

        output.TagName = "div";
        output.TagMode = TagMode.StartTagAndEndTag;
        output.Attributes.SetAttribute("class", "table-responsive");
        var html = $@"
{childHtml}
<table
    id=""{tableId}""
    class=""table table-hover table-bordered table-nowrap align-middle mb-0 vlz-data-table""
    data-ajax-url=""{AjaxUrl}""
    data-method=""{NormalizedMethod}""
    data-page-size=""{PageSize}""
    data-enable-sorting=""{EnableSorting.ToString().ToLowerInvariant()}""
    data-enable-pagination=""{EnablePagination.ToString().ToLowerInvariant()}""
    data-enable-column-selection=""{EnableColumnSelection.ToString().ToLowerInvariant()}""
    data-filter-form-id=""{FilterFormId}""
    data-show-filter-row=""{ShowFilterRow.ToString().ToLowerInvariant()}"">
  <thead></thead>
  <tbody>
    <tr><td colspan=""1"" class=""text-center py-4"">در حال بارگذاری…</td></tr>
  </tbody>
</table>";

        // Add initialization script
        html += $@"
<script>
  document.addEventListener('DOMContentLoaded', function() {{
    if (window.VlzTable) {{
      window.VlzTable.initTable(document.getElementById('{tableId}'));
    }} else {{
      console.error('VlzTable script not loaded. Make sure to include vlz-table.js in your page.');
    }}
  }});
</script>";

        output.Content.SetHtmlContent(html);
    }
}


// ───────────────────────────────────────── P O C O ─────────────────────────────────────────────────────
public sealed class VlzTableColumn
{
    public string  Name           { get; init; } = string.Empty;
    public string  Title          { get; init; } = string.Empty;
    public string? CssClass       { get; init; }
    public bool    Filterable     { get; init; }
    public string? FilterType     { get; init; }
    public string? TemplateId     { get; init; }
    public string? TemplateContent { get; init; }
    public bool    Sortable       { get; init; } = true;
    public string? Width      { get; init; }
    public bool    Visible    { get; init; } = true;
    public string? DataType   { get; init; } = "text";
    public string? Format     { get; init; }
    public bool    Searchable { get; init; } = true;
    public string? DefaultValue { get; init; }
    public int     Order      { get; init; } = 0; // Column display order
    public string? Tooltip    { get; init; }
    public bool    Fixed      { get; init; }
    
    [JsonIgnore]
    public Dictionary<string, string> CustomAttributes { get; init; } = new();
}
